File: Assets/Import_out/Uzil/API/Invoker.cs
using System.Collections.Generic;

using UnityEngine;

using Uzil;
using Uzil.Lua;

using UzInvoker = Uzil.Invoker;

namespace UZAPI {

public class Invoker {

	/*======================================Constructor==========================================*/

	/*=====================================Static Members========================================*/

	/*=====================================Static Funciton=======================================*/

	/*=========================================Members===========================================*/

	private static Dictionary<InvokeTask, int> _task2cbID = new Dictionary<InvokeTask, int>();

	/*========================================Components=========================================*/

	/*==========================================Event============================================*/

	/*======================================Unity Function=======================================*/

	/*========================================Interface==========================================*/

	/** Delay呼叫 */
	public static void Once (string instName, int callbackID, float time, string tags = null) {
		
		UzInvoker invoker = UzInvoker.Inst(instName);

		// 註冊任務
		InvokeTask_Delay task = null;
		task = invoker.Once(() => {
			
			// 移除任務
			invoker.Cancel(task);
			
			// 呼叫LuaCB
			UZAPI.Callback.CallLua_cs(callbackID);

		}, time);

		// 設置 任務 標籤
		List<string> tagList = DictSO.List<string>(tags);
		if (tagList != null) {
			foreach (string tag in tagList) {
				task.Tag(tag);
			}
		}

		// 註冊 任務 與 CBID 對應
		Invoker._task2cbID.Add(task, callbackID);
	}

	/** 取消 */
	public static void Cancel (string instName, string tags) {
		
		UzInvoker inst = UzInvoker.Inst(instName);

		List<string> tagList = DictSO.List<string>(tags);
		if (tagList != null) {
			foreach (string tag in tagList) {
				List<InvokeTask_Delay> toRm = inst.Cancel(tag);
				foreach (InvokeTask_Delay task in toRm) {
					Invoker.removeCallbackID(task);
				}
			}
		}
	}

	/** 清空 */
	public static void Clear (string instName) {
		UzInvoker.Inst(instName).Clear();
		InvokerUpdate.Inst(instName).Clear();
	}

	/** 加入呼叫佇列 */
	public static void Queue (string instName, int callbackID) {
		InvokerSerial queue = InvokerSerial.Inst(instName);
		//註冊事件
		queue.Set(() => {
			UZAPI.Callback.CallLua_cs(callbackID);
		});
	}

	/** 設定佇列最後一個呼叫 */
	public static void QueueFinal (string instName, int callbackID) {
		InvokerSerial queue = InvokerSerial.Inst(instName);
		// 註冊事件
		queue.onEnd.Add(() => {
			UZAPI.Callback.CallLua_cs(callbackID);
		});
	}

	/** 呼叫 */
	public static void QueueCall (string instName) {
		InvokerSerial queue = InvokerSerial.Inst(instName);
		queue.DoAll();
	}

	/** 每幀 */
	public static void Update (string instName, int callbackID, string tags) {
		InvokerUpdate inst = InvokerUpdate.Inst(instName);

		// 註冊事件
		InvokeTask task = inst.Add(()=>{
			UZAPI.Callback.CallLua_cs_arg(callbackID, TimeUtil.deltaTime);
		});

		List<string> tagList = DictSO.List<string>(tags);
		if (tagList != null) {
			foreach (string tag in tagList) {
				task.Tag(tag);
			}
		}

		Invoker._task2cbID.Add(task, callbackID);
	}

	/** 取消每幀 */
	public static void UpdateCancel (string instName, string tags) {
		InvokerUpdate inst = InvokerUpdate.Inst(instName);

		List<string> tagList = DictSO.List<string>(tags);
		if (tagList != null) {
			foreach (string tag in tagList) {
				List<InvokeTask_Priority> toRm = inst.Cancel(tag);
				foreach (InvokeTask_Priority task in toRm) {
					Invoker.removeCallbackID(task);
				}
			}
		}
	}

	private static void removeCallbackID (InvokeTask task) {
		if (Invoker._task2cbID.ContainsKey(task) == false) return;

		int callbackID = Invoker._task2cbID[task];

		Invoker._task2cbID.Remove(task);
		
		Callback.Remove(callbackID);
	}

	/*=====================================Public Function=======================================*/

	/*===================================Protected Function======================================*/

	/*====================================Private Function=======================================*/

}

}

File: AppData/Mods/BUILT-IN/_lua/LuaScript/Uzil/Invoker.lua
local this = {};

--[[ 取得實例 ]] -- table
function this.inst(key)

	local instance = {};

	instance.key = key;

	-- [[ 呼叫 ]]
	instance.once = function (func, time, tagOrTags) 
		return this.once(instance.key, func, time, tagOrTags);
	end

	-- [[ 取消 ]]
	instance.cancel = function (tagOrTags) 
		this.cancel(instance.key, tagOrTags);
	end

	--[[ 清空 ]]
	instance.clear = function ()
		this.clear(instance.key);
	end

	-- [[ 呼叫 每幀 ]]
	instance.update = function (func, tagOrTags)
		this.update(instance.key, func, tagOrTags);
	end

	-- [[ 取消 每幀 ]]
	instance.updateCancel = function (tagOrTags) 
		this.updateCancel(instance.key, tagOrTags);
	end


	return instance;
end

--[[ 呼叫 ]]
function this.once (instID, func, time, tagOrTags)
	-- 確保 tags 為 複數
	local tags = Util.oneToArray(tagOrTags);
	-- 若時間未指定 則 預設0
	if (time == nil) then time = 0 end;
	-- 取得CBID
	local cbid = Callback.newID(nil, func, "invoker.once");
	-- 於 C#層 註冊 單次呼叫
	UZAPI.Invoker.Once(instID, cbid, time, Json.encode(tags));

	return cbid;
end

--[[ 取消 ]]
function this.cancel (instID, tagOrTags)
	local tags = Util.oneToArray(tagOrTags);
	UZAPI.Invoker.Cancel(instID, Json.encode(tags));
end

--[[ 清空 ]]
function this.clear (instID)
	UZAPI.Invoker.Clear(instID);
end

--[[ 呼叫 每幀 ]]
function this.update (instID, func, tagOrTags)
	local tags = Util.oneToArray(tagOrTags);

	local keyBase = "_invokerUpdate."..tostring(os.time());
	local key = keyBase;
	local fix = 1;

	while (Callback.isExist(key)) do
		key = keyBase.."."..tostring(fix);
		fix = fix + 1;
	end

	local cbid = Callback.getID(key, func);

	UZAPI.Invoker.Update(instID, cbid, Json.encode(tags));
end

--[[ 取消 每幀 ]]
function this.updateCancel (instID, tagOrTags)
	local tags = Util.oneToArray(tagOrTags);
	UZAPI.Invoker.UpdateCancel(instID, Json.encode(tags));
end

return this;